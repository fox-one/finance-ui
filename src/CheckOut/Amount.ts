import {
  Component,
  Vue,
  Prop
} from 'vue-property-decorator';
import classnames from '@utils/classnames';
import { transfer } from '@utils/checkout';
import { VLayout, VImg } from 'vuetify/lib';
/* import types */
import type { CreateElement, VNode } from 'vue';
import type { ChannelType, Fiat } from './CheckOut';

@Component
export class Amount extends Vue {
  @Prop({ type: String, default: '' }) private icon!: string;
  @Prop({ type: [Number, String], default: 0 }) private amount!: number | string;
  @Prop({ type: [Number, String], default: 0 }) private discount!: number | string;
  @Prop({ type: Number, default: 1 }) private quantity!: number;
  @Prop({ type: String, default: 'CNY' }) private symbol!: string;
  @Prop({ type: Object, default: () => ({}) }) private channel!: ChannelType;
  @Prop({ type: Object, default: () => ({ price_ratio: 1, symbol: '$' }) }) private fiat!: Fiat;

  private get actualAmount() {
    return transfer(
      1,
      this.channel.price_ratio,
      (+this.amount) - (+this.discount),
      this.quantity,
      this.channel.percision ?? 2
    );
  }

  public render(h: CreateElement): VNode {
    const classes = classnames('checkout-amount');
    return h(
      'div',
      {
        staticClass: `text--primary ${classes()}`
      },
      [
        h(
          VLayout,
          {
            attrs: {
              'align-center': true,
              'justify-center': true
            }
          },
          [
            h(
              'div',
              { staticClass: 'mr-3' },
              [
                h(
                  VImg,
                  {
                    props: {
                      width: '38',
                      height: '38',
                      src: this.icon
                    }
                  }
                )
              ]
            ),
            h(
              'div',
              [
                h(
                  'div',
                  { staticClass: 'font-weight-medium' },
                  [
                    h(
                      'span',
                      { staticClass: classes('major') },
                      `${this.actualAmount}`
                    ),
                    h(
                      'span',
                      { staticClass: classes('symbol') },
                      this.symbol
                    )
                  ]
                ),
                this.discount ? h(
                  'div',
                  {
                    staticClass: `text-center ${classes('original')}`
                  },
                  [
                    h(
                      'span',
                      `${this.amount}`
                    ),
                    h(
                      'span',
                      this.symbol
                    )
                  ]
                ) : null,
                h(
                  'div',
                  { staticClass: 'text-subtitle-2 text--secondary' },
                  `â‰ˆ ${this.fiat.symbol}${+this.amount / this.fiat.price_ratio}`
                )
              ]
            )
          ]
        )
      ]
    );
  }
}

export default Amount;