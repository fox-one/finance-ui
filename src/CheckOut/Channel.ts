import {
  Component,
  Vue,
  Prop
} from 'vue-property-decorator';
import checkIcon from '@assets/images/common/check.svg';
/* import types */
import type { CreateElement, VNode } from 'vue';
import type { ChannelType } from './CheckOut';

@Component
export class Channel extends Vue {
  @Prop({ type: Object, default: () => ({}) }) private channel!: ChannelType;
  @Prop({ type: Boolean, default: false }) private selected!: boolean;

  private loading = false;

  private get disabled() {
    return this.loading;
  }

  public render(h: CreateElement): VNode {
    return h(
      'v-list-item',
      {
        props: {
          ripple: false
        },
        on: {
          click: () => this.$emit('select', this.channel)
        }
      },
      [
        h(
          'v-list-item-avatar',
          {
            props: {
              size: '28'
            }
          },
          [
            h(
              'v-img',
              {
                props: {
                  src: this.channel.channel_icon || this.channel.asset_icon
                }
              }
            )
          ]
        ),
        h(
          'v-list-item-content',
          this.channel.channel_name
        ),
        this.selected ? h(
          'v-list-item-icon',
          [
            h(
              'v-img',
              {
                props: {
                  small: true,
                  color: 'primary',
                  src: checkIcon
                }
              }
            )
          ]
        ) : null
      ]
    );
  }
}

export default Channel;