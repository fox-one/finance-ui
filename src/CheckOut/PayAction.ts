import {
  Component,
  Vue,
  Prop
} from 'vue-property-decorator';
import { $t } from '@utils/t';
/* import types */
import type { CreateElement, VNode } from 'vue';

@Component
export class PayAction extends Vue {
  @Prop({ type: Function, default: () => {} }) private onPay!: () => any;

  private loading = false;

  private get disabled() {
    return this.loading;
  }

  private async handlePay() {
    this.$emit('pay');
    this.loading = true;
    try {
      await this.onPay?.();
    } catch (error) {
      this.$emit('error', error);
    }
    this.loading = false;
  }

  public render(h: CreateElement): VNode {
    return h(
      'div',
      [
        h(
          'f-button',
          {
            props: {
              block: true,
              type: 'primary',
              disabled: this.disabled,
              loading: this.loading
            },
            on: {
              click: () => this.handlePay()
            }
          },
          $t(this, 'pay')
        )
      ]
    );
  }
}

export default PayAction;