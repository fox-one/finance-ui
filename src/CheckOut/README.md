# FCheckOut

## Example

```tsx
import { FCheckOut } from './FCheckOut';

<f-check-out />
```
