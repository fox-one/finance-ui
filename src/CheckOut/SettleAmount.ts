import {
  Component,
  Vue,
  Prop,
  PropSync,
  Model
} from 'vue-property-decorator';
import classnames from '@utils/classnames';
import Bridge from '@foxone/mixin-sdk-jsbridge';
/* import types */
import type { CreateElement, VNode } from 'vue';

@Component
export class SettleAmount extends Vue {
  @Prop({ type: String, default: '' }) private title!: string;
  @Prop({ type: String, default: '' }) private subtitle!: string;
  @Prop({ type: String, default: '' }) private client_id!: string;
  @Prop({ type: String, default: '' }) private className!: string;

  @Model('change', { type: Boolean, default: false }) private show!: boolean;

  private bridge: undefined | Bridge;

  private loading = false;

  public created() {
    if (this.client_id) {
      this.bridge = new Bridge({
        client_id: this.client_id
      });
    }
  }

  public render(h: CreateElement): VNode {
    const classes = classnames({ prefix: 'prefix' });
    return h(
      'div',
      {
        staticClass: classes('checkout'),
        class: this.className
      },
      [
        h(
          'f-bottom-sheet',
          {
            props: {
              value: this.show
            },
            on: {
              change: (val: any) => {
                this.$emit('change', val);
              }
            }
          },
          [
            !this.loading
              ? [
                this.title ? h(
                  'template',
                  {
                    slot: 'title',
                  },
                  [
                    h('h4', { staticClass: 'f-greyscale-1' }, this.title)
                  ]
                ) : null,
                this.subtitle ? h(
                  'template',
                  {
                    slot: 'subheader',
                  },
                  [
                    h('h5', { staticClass: 'f-greyscale-3 f-caption subtitle text-center' }, this.subtitle)
                  ]
                ) : null,
                h('div', 'test msg123')
              ]
              : h(
                'div',
                {
                  staticClass: 'ma-10 text-center'
                },
                'loading'
              )
          ]
        )
      ]
    );
  }
}

export default SettleAmount;