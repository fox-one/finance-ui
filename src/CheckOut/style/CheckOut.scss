.prefix-checkout {
  display: block;
  color: green;
}
