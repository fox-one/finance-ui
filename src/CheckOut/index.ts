import CheckOut from './CheckOut';

export { CheckOut } from './CheckOut';

export default CheckOut;
