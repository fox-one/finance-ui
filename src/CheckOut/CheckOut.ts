import {
  Component,
  Vue,
  Prop,
  Model
} from 'vue-property-decorator';
import classnames from '@utils/classnames';
import Amount from './Amount';
import Channel from './Channel';
import PayAction from './PayAction';
import { FBottomSheet } from '@foxone/uikit/src/components/FBottomSheet';
/* import types */
import type { CreateElement, VNode } from 'vue';

export interface ChannelType {
  channel: 'mixin' | 'wechat' | 'alipay';
  channel_name: string;
  channel_icon: string;
  asset_id: string;
  asset_icon: string;
  symbol: string;
  percision?: number;
  price_ratio: number;
  [props: string]: any;
}

export interface Fiat {
  symbol: string;
  price_ratio: number;
}

@Component({
  components: {
    Amount,
    Channel,
    PayAction
  }
})
export class CheckOut extends Vue {
  // Data relative
  @Prop({ type: Array, default: () => [] }) private channels!: ChannelType[];
  @Prop({ type: [Number, String], default: 0 }) private amount!: number | string;
  @Prop({ type: [Number, String], default: 0 }) private discount!: number | string;
  @Prop({ type: Number, default: 1 }) private quantity!: number;
  @Prop({ type: Object, default: () => ({ price_ratio: 1, symbol: '$' }) }) private fiat!: Fiat;

  // UI relative
  @Prop({ type: String, default: '' }) private title!: string;
  @Prop({ type: String, default: '' }) private subtitle!: string;
  @Prop({ type: String, default: 'checkout' }) private prefixCls!: string;

  @Model('change', { type: Boolean, default: false }) private show!: boolean;


  private loading = false;

  private paychannel: ChannelType = {} as ChannelType;

  private isChannelSelected(channel: ChannelType) {
    return (
      this.paychannel?.channel === channel.channel &&
      this.paychannel.asset_id === channel.asset_id
    );
  }

  public created() {
    this.paychannel = this.channels?.[0];
  }

  public render(h: CreateElement): VNode {
    const classes = classnames(this.prefixCls);
    return h(
      'div',
      {
        staticClass: classes()
      },
      [
        h(
          FBottomSheet,
          {
            props: {
              value: this.show
            },
            attrs: {
              persistent: true
            },
            on: {
              change: (val: any) => {
                this.$emit('change', val);
              }
            }
          },
          [
            !this.loading
              ? [
                this.title ? h(
                  'template',
                  {
                    slot: 'title',
                  },
                  [
                    h('h4', { staticClass: 'f-greyscale-1' }, this.title)
                  ]
                ) : null,
                this.subtitle ? h(
                  'template',
                  {
                    slot: 'subheader',
                  },
                  [
                    h('h5', { staticClass: 'f-greyscale-3 f-caption subtitle text-center' }, this.subtitle)
                  ]
                ) : null,
                h(
                  'amount',
                  {
                    props: {
                      amount: this.amount,
                      discount: this.discount,
                      quantity: this.quantity,
                      symbol: this.paychannel.symbol,
                      icon: this.paychannel.asset_icon || this.paychannel.channel_icon,
                      channel: this.paychannel,
                      fiat: this.fiat
                    }
                  }
                ),
                this.channels?.map((channel, ind) => h(
                  'channel',
                  {
                    key: ind,
                    props: {
                      channel,
                      selected: this.isChannelSelected(channel)
                    },
                    on: {
                      select: (channel: ChannelType) => {
                        this.paychannel = channel;
                      }
                    }
                  }
                )),
                h(
                  'pay-action',
                  {
                    on: {
                      pay: () => {
                        console.info('pay action!');
                        this.$emit('pay');
                      }
                    }
                  }
                )
              ]
              : h(
                'div',
                {
                  staticClass: 'ma-10 text-center'
                },
                'loading'
              )
          ]
        )
      ]
    );
  }
}

export default CheckOut;