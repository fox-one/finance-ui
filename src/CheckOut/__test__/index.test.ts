import { shallowMount, mount } from '@vue/test-utils';
import CheckOut from '../';

describe('CheckOut', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(CheckOut);
    expect(wrapper).toMatchSnapshot();
  });
});
