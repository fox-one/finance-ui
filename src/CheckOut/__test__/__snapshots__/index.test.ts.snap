// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckOut renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {
    "hook:mounted": Array [
      Array [],
    ],
  },
  "_emittedByOrder": Array [
    Object {
      "args": Array [],
      "name": "hook:mounted",
    },
  ],
  "isFunctionalComponent": undefined,
}
`;
